fl4health/parameter_exchange/parameter_selection_criteria.py

```diff
@@ -1,13 +1,58 @@
 import math
-from typing import Dict, List, Tuple
+from functools import partial
+from typing import Callable, Dict, List, Optional, Tuple
 
 import torch
 import torch.nn as nn
 from flwr.common.typing import NDArrays
 from torch import Tensor
 
-
-# Selection criteria functions for selecting entire layers. Should be used with the DynamicLayerExchanger class.
+LayerSelectionFunction = Callable[[nn.Module, nn.Module], Tuple[NDArrays, List[str]]]
+
+
+class LayerSelectionFunctionConstructor:
+    def __init__(
+        self, norm_threshold: float, exchange_percentage: float, normalize: bool = True, select_drift_more: bool = True
+    ) -> None:
+        """
+        This class leverages functools.partial to construct layer selection functions,
+        which are meant to be used by the DynamicLayerExchanger class.
+
+        Args:
+            norm_threshold (float): A nonnegative real number used to select those
+                layers whose drift in l2 norm exceeds (or falls short of) it.
+            exchange_percentage (float): Indicates the percentage of layers that are selected.
+            normalize (bool, optional): Indicates whether when calculating the norm of a layer,
+                we also divide by the number of parameters in that layer. Defaults to True.
+            select_drift_more (bool, optional): Indicates whether layers with larger
+                drift norm are selected. Defaults to True.
+        """
+        assert 0 < exchange_percentage <= 1
+        assert 0 < norm_threshold
+        self.norm_threshold = norm_threshold
+        self.exchange_percentage = exchange_percentage
+        self.normalize = normalize
+        self.select_drift_more = select_drift_more
+
+    def select_by_threshold(self) -> LayerSelectionFunction:
+        return partial(
+            select_layers_by_threshold,
+            self.norm_threshold,
+            self.normalize,
+            self.select_drift_more,
+        )
+
+    def select_by_percentage(self) -> LayerSelectionFunction:
+        return partial(
+            select_layers_by_percentage,
+            self.exchange_percentage,
+            self.normalize,
+            self.select_drift_more,
+        )
+
+
+# Selection criteria functions for selecting entire layers. Intended to be used
+# by the DynamicLayerExchanger class via the LayerSelectionFunctionConstructor class.
 def _calculate_drift_norm(t1: torch.Tensor, t2: torch.Tensor, normalize: bool) -> float:
```

Comment:
I might be wrong, but I think we can drop the underscore prefix here? I know the underscore is used to indicate "private" class functions. I'm not sure if a similar convention holds for stand-alone functions?

Reply:
I think the same convention holds for stand-alone functions as well. For example, if you do a star import on the module, python will ignore the functions whose names begin with an underscore.

Reply:
I didn't know that! Thanks for the info.